### ISSUE TITLE
Not able to spawn Airbyte connectors on Jitsu running on Kubernetes using DinD

### ISSUE THREAD

**Issue by shubham19may** [jitsucom/jitsu#863]
## Environment

- **Deploy method**: `jitsucom/jitsu docker`
- **Jitsu version**:  `latest`
- **OS**: `Linux`
- **Cloud Provider**: `AWS EKS`
- **Docker Params (if deployed with docker)**: Mounting all required file paths including airbyte & docker.sock

## Description
Trying to deploy Jitsu on kubernetes with docker as DinD mode mentioned in the issue https://github.com/jitsucom/jitsu/issues/545. 
The log says ❌ Airbyte integration is disabled: failed to inspect current docker container by containerID [jitsu]: Error: No such container: jitsu

## Steps to reproduce

This is the configuration i have used to create a run Jitsu on K8s.
```
apiVersion: apps/v1
kind: Deployment
metadata:
    name: jitsu
spec:
    selector:
        matchLabels:
            app: jitsu
    template:
        metadata:
            labels:
                app: jitsu
        spec:
            hostname: jitsu
            volumes:
                - name: jitsu
                  persistentVolumeClaim:
                      claimName: jitsu-claim
                - name: docker-graph-storage 
                  emptyDir: {}
            containers:
                - name: jitsu
                  image: jitsucom/jitsu:latest
                  imagePullPolicy: IfNotPresent
                  env:
                      - name: CONFIGURATOR_ADMIN_TOKEN
                        value: kwaTnRNzdgMZ5gqDUAf3vNupvKw8Q9yY
                      - name: SERVER_ADMIN_TOKEN
                        value: TKNu4iPuJcvhsdfX9ukFqF6VhXmtpc4j
                      - name: UI_AUTH_ACCESS_SECRET
                        value: 8ydWLwpRWnHnKPMFq24a2iBZcMoyMfYL
                      - name: UI_AUTH_REFRESH_SECRET
                        value: EtahBGriN99tGt5EXQtYDBtoiMVnyWmA
                      - name: REDIS_URL
                        value: redis://redis:15601
                      - name: DOCKER_HOST 
                        value: tcp://localhost:2375 
                  ports:
                      - containerPort: 8000
                - name: dind-daemon 
                  image: docker:dind
                  command: ["dockerd", "--host", "tcp://127.0.0.1:2375"]
                  securityContext: 
                      privileged: true 
                  volumeMounts: 
                    - mountPath: /home/configurator/data/logs
                      name: docker-graph-storage
                    - mountPath: /home/eventnative/data/logs/events
                      name: docker-graph-storage
                    - mountPath: /home/eventnative/data/logs
                      name: docker-graph-storage
                    - mountPath: /home/eventnative/data/airbyte
                      name: docker-graph-storage
                    - mountPath: "/var/run/docker.sock"
                      name: docker-graph-storage
                        
                
            initContainers:
                - name: init-jitsu
                  image: jitsucom/jitsu:latest
                  command: ['sh', '-c', 'set -x', 'chmod -R 777 /home/eventnative', 'chmod -R 777 /home/configurator']
                  volumeMounts:
                      - mountPath: /home/configurator/data/logs
                        name: docker-graph-storage
                      - mountPath: /home/eventnative/data/logs/events
                        name: docker-graph-storage
                      - mountPath: /home/eventnative/data/logs
                        name: docker-graph-storage
                      - mountPath: /home/eventnative/data/airbyte
                        name: docker-graph-storage
                      - mountPath: "/var/run/docker.sock"
                        name: docker-graph-storage
```

## Expected behavior

Expected behaviour is to successfully run Airbyte connectors on kubernetes in DinD or DooD modes.

## Actual behavior

Currently when Jitsu starts and i create a simple file (airbyte) source. It throws following error in logs : 

```
2022-03-16 20:33:06 [ERROR]: source-file-0.2.9-6b1b8385-2f21-400c-8c9a-1909cc2b4136 airbyte runner closing failed. Killing. Closing error: exit status 1
```

And when i press next after filling details of File connector, it throws following error : 

```
2022-03-16 20:33:49 [ERROR]: Error executing airbyte check:
	{"type": "LOG", "log": {"level": "FATAL", "message": "[Errno 2] No such file or directory: '/tmp/airbyte/7cda4ceffc75441495f5e826c23ccb89/8fda8740c9374561b9f9fbf486229556.json'\nTraceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 13, in <module>\n    launch(source, sys.argv[1:])\n  File \"/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py\", line 127, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py\", line 80, in run\n    self.configure_sentry(source_spec.connectionSpecification, parsed_args)\n  File \"/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py\", line 67, in configure_sentry\n    config = self.source.read_config(parsed_args.config)\n  File \"/usr/local/lib/python3.7/site-packages/airbyte_cdk/connector.py\", line 42, in read_config\n    with open(config_path, \"r\") as file:\nFileNotFoundError: [Errno 2] No such file or directory: '/tmp/airbyte/7cda4ceffc75441495f5e826c23ccb89/8fda8740c9374561b9f9fbf486229556.json'"}}
```

**Comment by viggin543** [comment]
@shubham19may 
Don't kill the messenger, It's only my personal opinion. 

I think that DnD and k8s are not things you should run in prod, for a myriad of reasons.

We installed Airbyte on k8s as is.
It's a much cleaner approach IMHO.

**Comment by shubham19may** [comment]
> @shubham19may Don't kill the messenger, It's only my personal opinion.
> 
> I think that DnD on k8s is not something you should run in prod, for a myriad of reasons.
> 
> We installed Airbyte on k8s as is. It's a much cleaner approach IMHO.

@viggin543 so basically you are only using Airbyte on K8s rather than Jitsu?

And you are suggesting we should only use Airbyte because its more clean?

**Comment by viggin543** [comment]
@shubham19may clean === safe, monitorable, scalble etc..

**Comment by absorbb** [comment]
The fix was committed to master. We are going to give it day or two of extra testing before releasing new hotfix image.

Also i've noticed few mistakes in config example you have provided. E.g. all volumes was mounted to dind-daemon container instead of jitsu. init-jitsu command wouldn't  run anything except of `set -x`

I suggest config example with fixes and simplifications, and explanation how jitsu - DinD - airbyte schema works:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: jitsu
  labels:
    app: jitsu
spec:
  selector:
    matchLabels:
      app: jitsu
  template:
    metadata:
      name: jitsu
      labels:
        app: jitsu
    spec:
      hostname: jitsu
      volumes:
        - name: server-logs
          hostPath:
            path: /data/jitsu/server-logs
            type: DirectoryOrCreate
        - name: configutator-logs
          hostPath:
            path: /data/jitsu/configutator-logs
            type: DirectoryOrCreate
        - name: jitsu-workspace
          emptyDir: {}
        - name: docker-graph-storage 
          emptyDir: {}
      containers:
        - name: jitsu
          image: jitsucom/jitsu:latest
          imagePullPolicy: IfNotPresent
          env:
            - name: CONFIGURATOR_ADMIN_TOKEN
              value: admin_token
            - name: SERVER_ADMIN_TOKEN
              value: admin_token
            - name: REDIS_URL
              value: redis://host.minikube.internal:6379
            - name: DOCKER_HOST
              value: tcp://localhost:2375
            - name: SERVER_VOLUMES_WORKSPACE
              value: /tmp/airbyte/
          ports:
            - containerPort: 8000
          volumeMounts:
            - mountPath: /home/configurator/data/logs
              name: configutator-logs
            - mountPath: /home/eventnative/data/logs
              name: server-logs
            - mountPath: /home/eventnative/data/airbyte
              name: jitsu-workspace

        - name: dind-daemon
          image: docker:dind
          command: [ "dockerd", "--host", "tcp://127.0.0.1:2375" ]
          securityContext:
            privileged: true
          volumeMounts:
            - name: docker-graph-storage
              mountPath: /var/lib/docker
            - name: jitsu-workspace
              mountPath: /tmp/airbyte/

      initContainers:
        - name: init-jitsu
          image: jitsucom/jitsu:latest
          command: ["/bin/bash"]
          args: ['-c', 'sudo chmod -R 777 /home/eventnative; sudo chmod -R 777 /home/configurator']
          volumeMounts:
            - mountPath: /home/configurator/data/logs
              name: configutator-logs
            - mountPath: /home/eventnative/data/logs
              name: server-logs
```

### jitsu -> airbyte bridge through Docker-in-Docker

- Jitsu configured with SERVER_VOLUMES_WORKSPACE=/tmp/airbyte/
- Before Jitsu runs Airbyte source it generates config files and put them into `/home/eventnative/data/airbyte` directory
- `/home/eventnative/data/airbyte` directory mapped to **jitsu-workspace** named volume
- When Jitsu runs Airbyte source it uses command like `docker run -v SERVER_VOLUMES_WORKSPACE:/tmp/airbyte/ ... --config /tmp/airbyte/config.json` which translate to `docker run -v /tmp/airbyte/:/tmp/airbyte/ ... --config /tmp/airbyte/config.json`
- Since jitsu container have DOCKER_HOST=tcp://localhost:2375 airbyte source image actually runs inside dind-daemon container
- dind-daemon container have `/tmp/airbyte/` path mapped as **jitsu-workspace** named volume
- that is how through  **jitsu-workspace** named volume jitsu and airbyte source manage to interchange files

We will update docs with that information soon.

**Comment by shubham19may** [comment]
@absorbb thank you for the fix. I appreciate the prompt response. Will try it out.

One more thing though, DinD mode is not recommended to run things in prod : 

http://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

Do you think DinD approach is good solution for concurrent accesses (Using multiple airbyte sources or same source multiple times) ?

**Comment by absorbb** [comment]
@shubham19may Thank you for the article. It is very interesting.

I think concerns explained in this article dont apply to Jitsu case.
Our use case is quite simple. We dont run multiple docker instances and don't share anything critical between them.
At max we have docker from kubernetes contatiner runtime
and one docker instance from dind-daemon container. And nothing that may affect stability of docker shared between them - we don't share images, caches,etc - only source config jsons.
I don't see reasons why running multiple airbyte sources may be a problem here.

Regarding using in prod. Right now we don't test Jitsu on any k8s setups. But we have plans to start doing that in the future. I guess you need to make a decision based on your testing results.

**Comment by viggin543** [comment]
@absorbb 
Making jitsu cloud-native ( Jitsu + k8s ) will greatly increase adoption, and can really explode in popularity.
A high-quality ( bitnami style ) helm chart will make running jitsu simple and safe.
It will allow one-click production-ready installation.  that can be configured and fine-tuned to every ORG needs.

There will always be people that would like to use a managed Jitsu regardless of how easy it is to run on your personal cloud.  So don't worry about that.

And cloud-native means that you don't use DnD.   You deploy stuff horizontally.
Airbyte has the same problems. Maybe they want it to be a bit hard to run it yourself...

Running Jitsu k8s on prod should be as easy as running a MongoDB, redis, or Postgres with regards to DevOps resources required to operate and install it.

**Comment by absorbb** [comment]
Fixed in [1.41.3](https://github.com/jitsucom/jitsu/releases/tag/v1.41.3)

@viggin543 
Thank you for sharing this. We started moving to this direction. We don't forget about your helm chart work either 👍
